### PR TITLE
Surface history of a ledger inscription

### PR DESCRIPTION
**Pull request by basilesimon**
This PR implements the logic on the frontend for the bubbling up of records' histories. As we've seen in #20, it's quite the process. The logic is namespaced and attemps to mimic #15.

Kapture 2021-07-08 at 18 44 10

---

A `Ledger` type is created in `gui/src/Ledger/` and contains:

* the Typescript types (e.g. describing the data our of QLDB and other shapes),
* the logic for dealing with API requests (e.g. listing the ledger records, fetching an item's history),
* a Svelte component handling the UI of each record,
* and the Svelte store logic.

The latter is particularly tricky and goes like this:

* `$ledgerData` is instantiated as a writable store which all parts of the app can access,
* it is populated on first load with the contents of the ledger, but this data does _not_ contain history info.
* The user can click a button to request an item's history, 
* which triggers an `update` of the store: the item's history is added _directly_ to the store.

### CODE REVIEW

File: gui/src/Ledger/index.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Show a record's history in the UI: fetch this history
+ * and add it to the `historyData` store
+ * @param entry object - the record as represented in QLDB
+ */
+export async function addHistoryTo(entry: LedgerEntry) {
+  const { sku } = entry;
+  const itemHistoryData = await fetchItemHistory(sku);
+
+  ledgerData.update(async (d: Promise<LedgerEntry[]>) => {
```
Comment (nikonikoniko) [inline gui/src/Ledger/index.ts:52]: this is a svelte state management syntax?  why does the lambda get passed a promise?

File: gui/src/Ledger/index.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    // let newStore;
+    let newStore = await d.then(data => {
```
Comment (nikonikoniko) [inline gui/src/Ledger/index.ts:54]: this whole function just stores extra info about a ledger item in the svelte store, right?

File: gui/src/Ledger/LedgerEntry.svelte [hunk 3]
```diff
@@ -1,6 +1,12 @@
 <script lang="ts">
-  import type { LedgerEntry } from './types';
+  import * as Ledger from './index';
+  import type { LedgerEntry, EntryHistory } from './index';
   export let entry: LedgerEntry;
+
+  let originalTX: null | EntryHistory = null;
+  $: if (entry.history) {
+    originalTX = Ledger.getOriginalTX(entry.history[0]);
```
Comment (nikonikoniko) [inline gui/src/Ledger/LedgerEntry.svelte:8]: should the `[0]` unpacking live in the `getOriginalTx` function?

File: gui/src/Ledger/index.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  thumb?: string;
+  history?: QLDBHistory[];
+};
+
+export type EntryHistory = {
+  originalTxDate: string;
+  originalTxTime: string;
+};
+
+export type QLDBHistory = {
```
Comment (nikonikoniko) [inline gui/src/Ledger/index.ts:18]: My naive feeling is that this should be `QLDBHistoryItem`, and `QLDBHistory` should be `QLDBHistoryItem[]`, but it's possible that's only my understanding of it

**Comment by nikonikoniko** [comment]
A fair question to ask is - do you want to require an extra step on the user's part to request the history?  It could also be sent as part of the api response on the original `list` step, too.

**Comment by basilesimon** [comment]
Yup, fair question. As you can see from the current implementation, my understanding of how it’s handled on the AWS side is that we need to jump through an extra hoop to access this (which consists of two requests).

Listing the ledger records, on the other hand, is dynamic and consists of one request performed by the API every time the client asks for it. 

Adding history ahead of document listing would slow the listing by the number of ledger records times the time it takes for these requests (and at the end of the day, SQL queries) to run.

**Comment by nikonikoniko** [comment]
The simple answer would be to paginate or otherwise restrict the requests to QLDB - would something like this be possible?  At a certain point a user will have to search through or filter their results anyways, as their collection grows.

File: gui/src/Ledger/LedgerEntry.svelte (hunk 3 above)
Comment (basilesimon) [reply gui/src/Ledger/LedgerEntry.svelte:8]: Yes that's a boo-boo indeed!

File: gui/src/Ledger/index.ts (hunk 4 above)
Comment (basilesimon) [reply gui/src/Ledger/index.ts:18]: Coming back to it now, what you say makes sense to me as well. Marking this as to do.

File: gui/src/Ledger/index.ts (hunk 2 above)
Comment (basilesimon) [reply gui/src/Ledger/index.ts:54]: yes!

File: gui/src/Ledger/index.ts (hunk 1 above)
Comment (basilesimon) [reply gui/src/Ledger/index.ts:52]: Yes this is clunky. Help needed on this async front, if I'm honest...

Comment (nikonikoniko) [reply gui/src/Ledger/index.ts:54]: Happy to go through it together! If it works it works

**Pull request merged**